low_level_platform/impl/src/lf_flexpret_support.c

```diff
@@ -0,0 +1,200 @@
+***************/
+
+/** Support file for Bare-metal FlexPRET platform.
+ *
+ *  @author{Shaokai Lin <shaokai@berkeley.edu>}
+ *  @author{Magnus Mæhlum <magnmaeh@stud.ntnu.no>}
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include "low_level_platform.h"
+
+/**
+ * Used to keep track of the number of nested critical sections.
+ *
+ * We should only disable interrupts when this is zero and we enter a critical section
+ * We should only  enable interrupts when we exit a critical section and this is zero
+ */
+static int critical_section_num_nested[FP_THREADS] = THREAD_ARRAY_INITIALIZER(0);
+
+/**
+ * @return 0 for success, or -1 for failure
+ */
+
+static volatile uint32_t last_time = 0;
+static volatile uint64_t epoch = 0;
+#define EPOCH_DURATION_NS (1ULL << 32)
+int _lf_clock_gettime(instant_t* t) {
+  uint32_t now = rdtime();
+  if (now < last_time) {
+    epoch += EPOCH_DURATION_NS;
+  }
+  *t = now + epoch;
+  last_time = now;
+  return 0;
+}
+
+int _lf_interruptable_sleep_until_locked(environment_t* env, instant_t wakeup_time) {
+  int ret = 0;
+  if (wakeup_time < 0) {
+    return ret;
+  }
+
+  // Enable interrupts and execute wait until instruction
+  lf_critical_section_exit(env);
+
+  // Wait until will stop sleep if interrupt occurs
+  fp_wait_until(wakeup_time);
+
+  if ((instant_t)rdtime64() < wakeup_time) {
+    // Interrupt occurred because we did not wait full wakeup_time
+    ret = -1;
+  }
+
+  lf_critical_section_enter(env);
+  return ret;
+}
+
+int lf_sleep(interval_t sleep_duration) {
+  // FIXME: Handle sleep durations exceeding 32bit
+  fp_delay_for(sleep_duration);
+  return 0;
+}
+
+/**
+ * Initialize the LF clock.
+ */
+void _lf_initialize_clock() {
+  // FlexPRET clock does not require any initialization
+}
+
+int lf_disable_interrupts_nested() {
+  // In the special case where this function is called during an interrupt
+  // subroutine (isr) it should have no effect
+  if ((read_csr(CSR_STATUS) & 0x04) == 0x04)
+    return 0;
+
+  uint32_t hartid = read_hartid();
+
+  fp_assert(critical_section_num_nested[hartid] >= 0, "Number of nested critical sections less than zero.");
+  if (critical_section_num_nested[hartid]++ == 0) {
+    fp_interrupt_disable();
+  }
+  return 0;
+}
+
+int lf_enable_interrupts_nested() {
+  // In the special case where this function is called during an interrupt
+  // subroutine (isr) it should have no effect
+  if ((read_csr(CSR_STATUS) & 0x04) == 0x04)
+    return 0;
+
+  uint32_t hartid = read_hartid();
+
+  if (--critical_section_num_nested[hartid] == 0) {
+    fp_interrupt_enable();
+  }
+  fp_assert(critical_section_num_nested[hartid] >= 0, "Number of nested critical sections less than zero.");
+  return 0;
+}
+
+/**
+ * Pause execution for a number of nanoseconds.
+ *
+ * @return 0 for success, or -1 for failure. In case of failure, errno will be
+ *  set appropriately (see `man 2 clock_nanosleep`).
+ */
+int lf_nanosleep(interval_t requested_time) {
```

Comment:
This can be implemented like lf_sleep above, no need for busy wait right?

Reply:
I ended up doing a `_lf_sleep_common` that is a catch-all